### PR TITLE
Redmine#7346: implement regex_replace

### PR DESCRIPTION
**Pull request by tzz**
See https://dev.cfengine.com/issues/7346

Implements `regex_replace` with no external dependencies.

Example:

```
bundle agent main
{
  vars:
      "test" string => "abcdefghij";
      "void" string => regex_replace("", $(test), "", "g");
      "void1" string => regex_replace(".", $(test), "", "g");
      "void2" string => regex_replace("", $(test), "x", "g");
      "allofitwithTEST" string => regex_replace(".*", $(test), "TEST", "");
      "nomatch" string => regex_replace("nonesuch", $(test), "TEST", "g");
      "dots_everywhere" string => regex_replace(".", $(test), "-DOT-", "g");
      "cap123" string => regex_replace("(...)", $(test), "[cap=$1]", "g");
      "simple_nocase" string => regex_replace("ABC", $(test), "ABC", "gi"); # case-insensitive
      "simple" string => regex_replace("abc", $(test), "ABC", "g");

  methods:
      "" usebundle => expected();
      "" usebundle => report();
}

bundle agent expected
{
  vars:
      "void" string => $(main.test);
      "void1" string => "";
      "void2" string => "xaxbxcxdxexfxgxhxixjx";
      "allofitwithTEST" string => "TEST";
      "nomatch" string => $(main.test);
      "dots_everywhere" string => "-DOT--DOT--DOT--DOT--DOT--DOT--DOT--DOT--DOT-";
      "cap123" string => "[cap=abc][cap=def][cap=ghi]j";
      "simple" string => "ABCdefghij";
      "simple_nocase" string => "ABCdefghij";

  methods:
      "" usebundle => report();
}

bundle agent report
{
  vars:
      "state" data => bundlestate("main");
      "state_string" string => storejson(state);

      "expected" data => bundlestate("expected");
      "expected_string" string => storejson(expected);

  reports:
      "ACTUAL: $(state_string)";
      "EXPECTED: $(expected_string)";
}
```

output:

```
R: ACTUAL: {
  "allofitwithTEST": "TEST",
  "cap123": "[cap=abc][cap=def][cap=ghi]j",
  "dots_everywhere": "-DOT--DOT--DOT--DOT--DOT--DOT--DOT--DOT--DOT--DOT-",
  "nomatch": "abcdefghij",
  "simple": "ABCdefghij",
  "simple_nocase": "ABCdefghij",
  "test": "abcdefghij",
  "void": "abcdefghij",
  "void1": "",
  "void2": "xaxbxcxdxexfxgxhxixjx"
}
R: EXPECTED: {
  "allofitwithTEST": "TEST",
  "cap123": "[cap=abc][cap=def][cap=ghi]j",
  "dots_everywhere": "-DOT--DOT--DOT--DOT--DOT--DOT--DOT--DOT--DOT-",
  "nomatch": "abcdefghij",
  "simple": "ABCdefghij",
  "simple_nocase": "ABCdefghij",
  "void": "abcdefghij",
  "void1": "",
  "void2": "xaxbxcxdxexfxgxhxixjx"
}
```

### CODE REVIEW

**Comment by nickanderson** [comment]
@kacfengine @jimis @estenberg anything in particular holding this up?

File: libutils/pcre_wrap.c [hunk 1]
```diff
@@ -0,0 +1,495 @@
...
+    pcre_wrap_substitute *rstruct;
+
+    /*
+     * Sanity check
+     */
+    if (NULL == replacement)
+    {
+        replacement = "";
+    }
+    int length = (int)strlen(replacement); // intentional
```
Comment (jimis) [inline libutils/pcre_wrap.c]: Intentional, but what's the intent? The comment should state it. Otherwise `size_t` is preferred.

File: libutils/pcre_wrap.c [hunk 2]
```diff
@@ -0,0 +1,495 @@
...
+
+    rstruct = (pcre_wrap_substitute *)calloc(1, sizeof(pcre_wrap_substitute));
```
Comment (jimis) [inline libutils/pcre_wrap.c]: Cast not needed.

File: libutils/pcre_wrap.c [hunk 3]
```diff
@@ -0,0 +1,495 @@
...
+    if (NULL == rstruct)
+    {
+        *errptr = PCRE_WRAP_ERR_NOMEM;
+        return NULL;
+    }
+
+    text = (char *)calloc(1, length + 1);
```
Comment (jimis) [inline libutils/pcre_wrap.c]: Same here with unnecessary casting of `(void *)` return value and a couple of other places as well.

File: libutils/pcre_wrap.c [hunk 4]
```diff
@@ -0,0 +1,495 @@
...
+        free(rstruct);
+        *errptr = PCRE_WRAP_ERR_NOMEM;
+        return NULL;
+    }
+
+    int textpos = 0;
+
+    if (trivialflag)
+    {
+        text = strncpy(text, replacement, length + 1);
```
Comment (jimis) [inline libutils/pcre_wrap.c]: `s/strncpy/memcpy/`

**Comment by jimis** [comment]
The code generally looks good, but I can't possibly follow with all the logic in here! :-) I would just trust the code if it came with a bunch of unit tests, trying a bunch of corner cases like empty strings, backreferences, quoting etc.

File: libutils/pcre_wrap.c (hunk 1 above)
Comment (tzz) [inline libutils/pcre_wrap.c]: OK, explained that we want to cap it to MAXINT

File: libutils/pcre_wrap.c (hunk 2 above)
Comment (tzz) [inline libutils/pcre_wrap.c]: OK; fixed.

File: libutils/pcre_wrap.c (hunk 3 above)
Comment (tzz) [inline libutils/pcre_wrap.c]: I think the original code I used for guidance did this for compatibility reasons.  Can we leave it or do you feel it must be fixed?

File: libutils/pcre_wrap.c (hunk 4 above)
Comment (tzz) [inline libutils/pcre_wrap.c]: OK; changed.

**Comment by tzz** [comment]
I think the acceptance test is pretty thorough.  Please suggest more tests to add there.

**Comment by tzz** [comment]
PR rebased and pushed.

File: libutils/pcre_wrap.c [hunk 5]
```diff
@@ -0,0 +1,495 @@
...
+
+ * This is a rewrite of PCRS with the same API and behavior.
+
+ */
+
+

+
+#include <string.h>
+#include <ctype.h>
```
Comment (jimis) [inline libutils/pcre_wrap.c]: Always include our `<platform.h>` on top for portability reasons, which BTW makes the standard C library includes here redundant, so you can remove them.

File: libutils/pcre_wrap.h [hunk 6]
```diff
@@ -0,0 +1,95 @@
+#ifndef PCRE_WRAP_H_INCLUDED
+#define PCRE_WRAP_H_INCLUDED
+
+
+
```
Comment (jimis) [inline libutils/pcre_wrap.h]: Include `<platform.h>` on top also here.

**Comment by jimis** [comment]
> I think the acceptance test is pretty thorough. Please suggest more tests to add there.

A couple of quoting (backslash) cases.

I still feel like unit testing would be best for `BufferSearchAndReplace()` and you can reuse `buffer_test.c` for that so it should be quite easy. But it's not a blocker since you have an acceptance test.

File: libutils/pcre_wrap.c (hunk 3 above)
Comment (jimis) [inline libutils/pcre_wrap.c]: It's not a style issue, [this cast is dangerous](http://c-faq.com/malloc/mallocnocast.html). Its roots are in the pre-ANSI C era, but today it has a negative effect. So yes, please do a quick search for "alloc" and remove the casts.

File: libutils/pcre_wrap.c [hunk 7]
```diff
@@ -0,0 +1,495 @@
...
+    pcre_wrap_substitute *rstruct;
+
+    /*
+     * Sanity check
+     */
+    if (NULL == replacement)
+    {
+        replacement = "";
+    }
+    int length = (int)strlen(replacement); // intentional: we want to use at most MAXINT bytes
```
Comment (jimis) [inline libutils/pcre_wrap.c]: If you want to use MAXINT bytes at most, cast is not the way to go. Because if `strlen()` returns greater number (i.e. if string size is 2GB) then `length` becomes negative. Extreme case, I know, but possible.

IMHO just change `length`, `textpos` and `sourcepos` to size_t and you are good to know. Except if you have something else in mind that I don't see.

File: tests/acceptance/01_vars/02_functions/regex_replace.cf [hunk 8]
```diff
@@ -0,0 +1,51 @@
...
+      # no matches, original intact
+      "nomatch" string => regex_replace("nonesuch", $(test), "TEST", "g");
+      # replace any character with -DOT-
+      "dots_everywhere" string => regex_replace(".", $(test), "-DOT-", "g");
+      # capture any three characters and replace with a backreference
+      "cap123" string => regex_replace("(...)", $(test), "[cap=$1]", "g");
+      # match ABC case-insensitive and replace with ABC
+      "simple_nocase" string => regex_replace("ABC", $(test), "ABC", "gi"); # case-insensitive
+      # match abc and replace with ABC
+      "simple" string => regex_replace("abc", $(test), "ABC", "g");
```
Comment (jimis) [inline tests/acceptance/01_vars/02_functions/regex_replace.cf]: Plus one test for source string empty. Plus one test for both source and pattern string empty.

Comment (jimis) [inline tests/acceptance/01_vars/02_functions/regex_replace.cf]: What about `\1 \2` style backreferencing?

File: libutils/pcre_wrap.c (hunk 5 above)
Comment (tzz) [inline libutils/pcre_wrap.c]: OK; done.

File: libutils/pcre_wrap.c (hunk 7 above)
Comment (tzz) [inline libutils/pcre_wrap.c]: OK; done.

File: libutils/pcre_wrap.c (hunk 3 above)
Comment (tzz) [inline libutils/pcre_wrap.c]: OK; done.

File: libutils/pcre_wrap.h (hunk 6 above)
Comment (tzz) [inline libutils/pcre_wrap.h]: OK; done.

File: tests/acceptance/01_vars/02_functions/regex_replace.cf (hunk 8 above)
Comment (tzz) [inline tests/acceptance/01_vars/02_functions/regex_replace.cf]: Empty string tests added.

The `\1` backreference style was not supported, only `$1`.  I added it in a followup commit.  It has its own tests.  When a numeric backreference is requested and no such capture was made, the code replaces it with an empty string.

**Comment by tzz** [comment]
I would prefer to let the CFEngine developers work on the C unit tests, I don't know that part of the core.  Please feel free to add the unit tests as you suggested after this is merged.

**Comment by jimis** [comment]
This looks very good, thanks @tzz!

**Comment by jimis** [comment]
trigger build

**Comment by jimis** [comment]
Does it pass "make check" in your platform? It failed all over the place in our CI:
```
CCLD   str_test
./.libs/libstr.a(buffer.o): In function `BufferSearchAndReplace':
../../libutils/buffer.c:471: undefined reference to `pcre_wrap_compile'
```
You probably need to modify the unit test makefiles.

**Comment by tzz** [comment]
Could you please do this as a followup commit?  If you prefer I do it, I won't be able to do it until the weekend.  Sorry, I'm just booked solid.

**Comment by jimis** [comment]
> Could you please do this as a followup commit? If you prefer I do it, I won't be able to do it until the weekend. Sorry, I'm just booked solid.

Allright, I'll try to steal some time!

**Comment by jimis** [comment]
Fix amended in pull request #2392. Closing!
